src/main.rs

```diff
@@ -220,38 +277,252 @@ impl Fixer {
+            }
+        }
+    }
+
+    // This is based on `symbolic_debuginfo::macho:MachObject::load_address`.
+    // We need to define it because `goblin` doesn't have an equivalent, and we
+    // use `goblin` rather than `symbolic_debuginfo` for Mach-O binaries.
+    fn object_load_address(macho: &mach::MachO) -> u64 {
+        for seg in macho.segments.iter() {
+            if let Ok(name) = seg.name() {
+                if name == "__TEXT" {
+                    return seg.vmaddr;
+                }
+            }
+        }
+        0
+    }
+
+    /// Construct a key for the `sym_func_addrs` hash map.
+    fn sym_func_key(file_name: &str, func_name: &str) -> String {
+        format!("{}:{}", file_name, func_name)
+    }
+
+    /// Is this filename within an archive? E.g. `libfoo.a(bar.o)` means that
+    /// `bar.o` is within the archive `libfoo.a`. If so, return the archive
+    /// name.
+    fn is_within_archive(file_name: &str) -> Option<&str> {
+        if let (Some(index), true) = (file_name.find(".a("), file_name.ends_with(")")) {
+            let ar_file_name = &file_name[..index + 2];
+            Some(ar_file_name)
+        } else {
+            None
+        }
+    }
+
+    /// Read the debug info from a file referenced by an OSO entry in a Macho-O
+    /// symbol table.
+    fn do_macho_oso(
+        sym_func_addrs: &SymFuncAddrs,
+        file_name: &str,
+        data: &[u8],
+        interner: &mut Interner,
+        func_infos: &mut Vec<FuncInfo>,
+    ) -> Result<(), String> {
+        // Although we use `goblin` to iterate through the symbol
+        // table, we use `symbolic` to read the debug info from the
+        // object/archive, because it's easier to use.
+        let archive = Archive::parse(&data)
+            .map_err(|e| format!("({:?}) parse `{}` referenced by", e, file_name))?;
+        let mut found_x86_64 = false;
+        for object in archive.objects() {
+            // Get the x86-64 object from the fat binary. (On Mac,
+            // Firefox is only available on x86-64.)
+            let object = object.map_err(|e| {
+                format!(
+                    "({:?}) parse fat binary entry in `{}` referenced by",
+                    e, file_name
+                )
+            })?;
+            if object.arch() != Arch::Amd64 {
+                continue;
+            }
+            found_x86_64 = true;
+            let debug_session = object
+                .debug_session()
+                .map_err(|_| format!("read debug info from `{}` referenced by", file_name))?;
+            FileInfo::add(
+                &sym_func_addrs,
+                file_name,
+                debug_session,
+                interner,
+                func_infos,
+            );
+        }
+        if !found_x86_64 {
+            return Err(format!(
+                "find x86-64 code in the fat binary `{}` referenced by",
+                file_name
+            ));
+        }
```

Comment:
This doesn't feel very idiomatic, maybe it would be better to split out the search function and do it in reverse:
1. Find and gather all the x86-64 objects
2. If the resulting set is empty return `Err()`...
3. ... otherwise add the symbols from the set and return `Ok()`

Reply:
There should only be on x86-64 object, so I made it break and set an `Option` when it is found.